File: C++ Version/Class 1 Sorting Algorithms/QuickSort.cpp
void quickSort(int* a, int left, int right) {
  if (left >= right) {
    return;
  }
  int pivot = left + (right - left) / 2;
  int index = left;
  swap(&a[pivot], &a[right]);
  for (int i = left; i < right; i++) {
    if (a[i] < a[right]) {
      swap(&a[index], &a[i]);
      index++;
    }
  }
  swap(&a[index], &a[right]);
  quickSort(a, left, index - 1);
  quickSort(a, index+1, right);
}

void swap(int* a, int* b){
  int temp;
  temp = *b;
  *b   = *a;
  *a   = temp;
}

File: Java Version/Sort Algorithm/Quick Sort.java
public Solution{
  public int[] quickSort(int[] array){
    if (array == null) {
      return array;
    }
    quickSort(array, 0, array.length - 1);
    return array;
  }

  public void quickSort(int[] array, int left, int right){
    if (left >= right) {
      return;
    }
    int pivotIdx = partition(array, left, right);
    quickSort(array, left, pivotIdx-1);
    quickSort(array, pivotIdx+1, right);
  }

  private int partition(int[] array, int left, int right){
    int pivotIdx = left + (int)(Math.random()*(right - left));
    int pivot    = array[pivotIdx];
    swap(array, pivotIdx, right);
    int leftBound = left;
    int rightBound = right - 1;
    while (leftBound <= rightBound) {
      if (array[leftBound] < pivot) {
        leftBound++;
      } else{
        swap(array, leftBound, rightBound);
        rightBound--;
      }
    }
    swap(array, leftBound, right);
    return leftBound;
  }
}
